File: packages/core/src/geo.js
const EARTH_RADIUS_KM = 6371.0088;
const toRad = (deg) => (deg * Math.PI) / 180;

function isPoint(p) {
  return (
    p != null &&
    typeof p.lat === "number" &&
    typeof p.lng === "number" &&
    Number.isFinite(p.lat) &&
    Number.isFinite(p.lng)
  );
}

/**
 * Great-circle distance in kilometres.
 *
 * Returns null — never 0 — when either point lacks coordinates. GPS is
 * genuinely unavailable sometimes (DB.md section 2 makes lat/lng nullable),
 * and a zero would rank an unlocatable recycler first.
 */
export function haversineKm(a, b) {
  if (!isPoint(a) || !isPoint(b)) return null;
  const dLat = toRad(b.lat - a.lat);
  const dLng = toRad(b.lng - a.lng);
  const s =
    Math.sin(dLat / 2) ** 2 +
    Math.cos(toRad(a.lat)) * Math.cos(toRad(b.lat)) * Math.sin(dLng / 2) ** 2;
  return 2 * EARTH_RADIUS_KM * Math.asin(Math.min(1, Math.sqrt(s)));
}

/**
 * Speed implied by the two geotags and two timestamps a record carries.
 * Detector D7 thresholds this at 80 km/h.
 *
 * The 0.01-hour floor mirrors the GREATEST(..., 0.01) in the SERVER.md
 * section 10 query, so the JavaScript and SQL forms of D7 agree.
 */
export function impliedKmph({ from, to, fromTs, toTs }) {
  const km = haversineKm(from, to);
  if (km === null) return null;
  const hours = (new Date(toTs).getTime() - new Date(fromTs).getTime()) / 3_600_000;
  if (!Number.isFinite(hours)) return null;
  return km / Math.max(hours, 0.01);
}

File: server/aiml/bhaav_aiml/geo.py
"""Mirrors @bhaav/core/geo so the JavaScript and Python forms of D7 agree.
Returns None — never 0 — when a coordinate is missing (DB.md section 2)."""

from __future__ import annotations
import math
from datetime import datetime

EARTH_RADIUS_KM = 6371.0088


def _is_point(p) -> bool:
    return (
        p is not None
        and p[0] is not None
        and p[1] is not None
        and math.isfinite(p[0])
        and math.isfinite(p[1])
    )


def haversine_km(a, b) -> float | None:
    if not _is_point(a) or not _is_point(b):
        return None
    lat1, lng1 = a
    lat2, lng2 = b
    d_lat = math.radians(lat2 - lat1)
    d_lng = math.radians(lng2 - lng1)
    s = (
        math.sin(d_lat / 2) ** 2
        + math.cos(math.radians(lat1)) * math.cos(math.radians(lat2)) * math.sin(d_lng / 2) ** 2
    )
    return 2 * EARTH_RADIUS_KM * math.asin(min(1.0, math.sqrt(s)))


def implied_kmph(a, b, ts_a: str, ts_b: str) -> float | None:
    km = haversine_km(a, b)
    if km is None:
        return None
    hours = (datetime.fromisoformat(ts_b) - datetime.fromisoformat(ts_a)).total_seconds() / 3600.0
    # 0.01-hour floor, matching SERVER.md section 10's GREATEST(..., 0.01).
    return km / max(hours, 0.01)
